ci-cd/github-actions.md

```diff
@@ -0,0 +1,218 @@
+2. The second one is the build and push action.
+
+In this example, let us set the push flag to `true` as we also want to push. We’ll then add a tag to specify to always go to the latest version. Lastly, we’ll echo the image digest to see what was pushed.
+
+To set up the workflow:
+
+1. Go to your repository in GitHub and then click **Actions** > **New workflow**.
+2. Click **set up a workflow yourself** and add the following content:
+
+First, we will name this workflow:
+
+```yaml
+name: CI to Docker Hub
+ ```
+
+Then, we will choose when we run this workflow. In our example, we are going to do it for every push against the main branch of our project:
+
+```yaml
+on:
+  push:
+    branches: [ main ]
+```
+
+Now, we need to specify what we actually want to happen within our action (what jobs), we are going to add our build one and select that it runs on the latest Ubuntu instances available:
+
+```yaml
+jobs:
+
+  build:
+    runs-on: ubuntu-latest
+```
+
+Now, we can add the steps required. The first one is to use our PAT and username to log into Docker Hub. The second is the Builder, the action  uses BuildKit under the hood through a simple Buildx action which we will also setup
+
+```yaml
+    Steps:
+
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_HUB_USERNAME }}
+          password: ${{ secrets.DOCKER_HUB_ACCESS_TOKEN }}
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Build and push
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          context: ./
+          file: ./Dockerfile
+          push: true
+          tags: ushamandya/simplewhale:latest
+
+      - name: Image digest
+        run: echo ${{ steps.docker_build.outputs.digest }}
+```
+
+Now, let the workflow run for the first time and then tweak the Dockerfile to make sure the CI is running and pushing the new image changes:
+
+![CI to Docker Hub](images/ci-to-hub.png){:width="500px"}
+
+## Optimizing the workflow
+
+Next, let’s look at how we can optimize the GitHub Actions workflow through build cache. This has two main advantages:
+
+1. Build cache reduces the build time as it will not have to re-download all of the images, and
+2. It also reduces the number of pulls we complete against Docker Hub. We need to make use of GitHub cache to make use of this.
+
+Let us set up a Builder with a build cache. First, we need to set up cache for the builder.  In this example, let us add the path and keys to store this under using GitHub cache for this.
+
+```yaml
+
+      - name: Cache Docker layers
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
```

Comment:
The `restore-keys` is different to `key`?
I just want to double check if this is correct. 😅 

Seems correct as it is used in https://github.com/metcalfc/docker-action-examples/blob/main/.github/workflows/release.yml#L44-L50 👍